File: src/main/java/com/cuupa/classificator/services/kb/semantic/dataExtraction/DateExtract.java
package com.cuupa.classificator.services.kb.semantic.dataExtraction;

import org.jetbrains.annotations.NotNull;

import java.util.regex.Pattern;

public class DateExtract extends Extract {

	private static Pattern pattern;

	public DateExtract(@NotNull String regex) {
		pattern = Pattern.compile(regex);
	}

	@Override
	public Pattern getPattern() {
		return pattern;
	}

	@NotNull
	@Override
	public String normalize(@NotNull String value) {
		String[] split = value.split("\\.");
		String day = split[0];
		String month = split[1];
		String year = split[2];
		
		if(day.length() == 1) {
			day = "0" + day;
		}
		
		if(month.length() == 1) {
			month = "0"+ month;
		}
		
		if(year.length() == 2) {
			year = "20" + year;
		}
		return day + "." + month + "." + year;

	}

}

File: src/main/java/com/cuupa/classificator/services/kb/semantic/dataExtraction/DateExtract.kt
package com.cuupa.classificator.services.kb.semantic.dataExtraction

import com.cuupa.classificator.constants.RegexConstants
import java.util.regex.Pattern

class DateExtract(regex: String) : Extract(Pattern.compile(regex, Pattern.CASE_INSENSITIVE)) {

    // TODO: normalize months in non numeric values
    override fun normalize(value: String): String {
        if (value.contains(".")) {
            val dateFields = value.split(RegexConstants.dotPattern)
            return when (dateFields.size) {
                3 -> normalizeDate(dateFields)
                2 -> parseAndNormalizeDate(dateFields)
                else -> value
            }
        }
        return value
    }

    private fun parseAndNormalizeDate(dateFields: List<String>): String {
        val monthAndYear = dateFields[1].split(" ")
        return fill(dateFields[0], Months.get(monthAndYear[0]), monthAndYear[1])
    }

    private fun fill(day: String, month: String, year: String): String {
        var day1 = day
        var month1 = month
        var year1 = year
        if (day1.length == 1) {
            day1 = "0$day1"
        }
        if (month1.length == 1) {
            month1 = "0$month1"
        }
        if (year1.length == 2) {
            year1 = "20$year1"
        }
        return "$day1.$month1.$year1"
    }

    private fun normalizeDate(dateFields: List<String>) = fill(dateFields[0], dateFields[1], dateFields[2])

    companion object {
        const val name = "[DATE]"
    }
}


